drivers/led/ncp5623c.c

```diff
@@ -0,0 +1,214 @@
+	const struct led_info *leds_info;
+};
+
+struct ncp5623c_data {
+	struct led_data dev_data;
+};
+
+static const struct led_info *ncp5623c_led_to_info(const struct ncp5623c_config *config,
+						   uint32_t led)
+{
+	if (led < config->num_leds) {
+		return &config->leds_info[led];
+	}
+
+	return NULL;
+}
+
+static int ncp5623c_get_info(const struct device *dev, uint32_t led, const struct led_info **info)
+{
+	const struct ncp5623c_config *config = dev->config;
+	const struct led_info *led_info = ncp5623c_led_to_info(config, led);
+
+	if (!led_info) {
+		return -EINVAL;
+	}
+
+	*info = led_info;
+
+	return 0;
+}
+
+static int ncp5623c_set_color(const struct device *dev, uint32_t led, uint8_t num_colors,
+			      const uint8_t *color)
+{
+	const struct ncp5623c_config *config = dev->config;
+	const struct led_info *led_info = ncp5623c_led_to_info(config, led);
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (num_colors != 3) {
+		LOG_ERR("%s: invalid number of colors: got=%d, expected=%d", dev->name, num_colors,
+			3);
+		return -ENOSUP;
+	}
+
+	uint8_t buf[6] = {0x70, NCP5623_LED_PWM0, 0x70, NCP5623_LED_PWM1, 0x70, NCP5623_LED_PWM2};
+
+	buf[1] = buf[1] | color[0] / 8;
+	buf[3] = buf[3] | color[1] / 8;
+	buf[5] = buf[5] | color[2] / 8;
+
+	return i2c_burst_write_dt(&config->bus, NCP5623_LED_CURRENT | NCP5623C_MAX_BRIGHTNESS, buf,
+				  sizeof(buf));
+}
+
+static int ncp5623c_set_brightness(const struct device *dev, uint32_t led, uint8_t value)
+{
+	const struct ncp5623c_config *config = dev->config;
+	const struct led_info *led_info = ncp5623c_led_to_info(config, led);
+	int ret = 0;
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (value > 100) {
+		return -EINVAL;
+	}
+
+	if (led_info->num_colors == 3) {
+		uint8_t colors[3];
+
+		colors[0] = value * 255 / 100;
+		colors[1] = colors[0];
+		colors[2] = colors[0];
+
+		ret = ncp5623c_set_color(dev, led, led_info->num_colors, colors);
+	} else {
+
+		/* Rescale 0..100 to 0..31 */
+		value = value * NCP5623C_MAX_BRIGHTNESS / 100;
+
+		ret = i2c_reg_write_byte_dt(&config->bus, led_channels[led] | value, 0x70);
+
+		if (ret < 0) {
+			LOG_ERR("LED write failed");
+		}
+	}
+
+	return ret;
+}
+
+static inline int ncp5623c_led_on(const struct device *dev, uint32_t led)
+{
+	return ncp5623c_set_brightness(dev, led, 100);
+}
+
+static inline int ncp5623c_led_off(const struct device *dev, uint32_t led)
+{
+	return ncp5623c_set_brightness(dev, led, 0);
+}
+
+static int ncp5623c_led_init(const struct device *dev)
+{
+	const struct ncp5623c_config *config = dev->config;
+	int channel_count = 0;
+
+	if (!i2c_is_ready_dt(&config->bus)) {
+		LOG_ERR("I2C device not ready");
+		return -ENODEV;
+	}
+
+	for (int i = 0; i < config->num_leds; i++) {
+		channel_count += ncp5623c_led_to_info(config, i)->num_colors;
+	}
+
+	if (channel_count > NCP5623C_CHANNEL_COUNT) {
+		LOG_ERR("%s: invalid number of channels %d (max %d)", dev->name, channel_count,
+			NCP5623C_CHANNEL_COUNT);
+		return -EINVAL;
+	}
```

Comment:
With this checks you could have the hardware configuration I was talking earlier: one dual-color LED and one single-color LED. It is maybe not what you want.